apis/v1alpha1/tcproute_types.go

```diff
@@ -19,38 +19,96 @@ import (
+// TCPRouteStatus defines the observed state of TCPRoute
+type TCPRouteStatus struct {
+	GatewayRefs []GatewayObjectReference `json:"gatewayRefs" protobuf:"bytes,1,rep,name=gatewayRefs"`
 }
 
 // +genclient
 // +kubebuilder:object:root=true
 
-// TcpRoute is the Schema for the tcproutes API
-type TcpRoute struct {
+// TCPRoute is the Schema for the tcproutes API
+type TCPRoute struct {
 	metav1.TypeMeta   `json:",inline" protobuf:"bytes,1,opt,name=typeMeta"`
 	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,2,opt,name=metadata"`
 
-	Spec   TcpRouteSpec   `json:"spec,omitempty" protobuf:"bytes,3,opt,name=spec"`
-	Status TcpRouteStatus `json:"status,omitempty" protobuf:"bytes,4,opt,name=status"`
+	Spec   TCPRouteSpec   `json:"spec,omitempty" protobuf:"bytes,3,opt,name=spec"`
+	Status TCPRouteStatus `json:"status,omitempty" protobuf:"bytes,4,opt,name=status"`
+}
+
+// TCPRouteRule is the configuration for a given rule.
+type TCPRouteRule struct {
+	// Match defines which connections match this rule.
+	//
+	// +optional
+	Match *TCPRouteMatch `json:"match" protobuf:"bytes,1,opt,name=match"`
+	// Action defines what happens to the connection.
+	//
+	// +optional
+	Action *TCPRouteAction `json:"action" protobuf:"bytes,2,opt,name=action"`
+}
+
+// TCPRouteAction is the action for a given rule.
+type TCPRouteAction struct {
+	// ForwardTo sends requests to the referenced object.  The
+	// resource may be "services" (omit or use the empty string for the
+	// group), or an implementation may support other resources (for
+	// example, resource "myroutetargets" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "services".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the TCPRoute will be true.
+	ForwardTo *ForwardToTarget `json:"forwardTo" protobuf:"bytes,1,opt,name=forwardTo"`
+
+	// ExtensionRef is an optional, implementation-specific extension to the
+	// "action" behavior.  The resource may be "configmaps" (use the empty
+	// string for the group) or an implementation-defined resource (for
+	// example, resource "myrouteactions" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "configmaps".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the TCPRoute will be true.
+	//
+	// Support: custom
+	//
+	// +optional
+	ExtensionRef *RouteActionExtensionObjectReference `json:"extensionRef" protobuf:"bytes,2,opt,name=extensionRef"`
+}
+
+// TCPRouteMatch defines the predicate used to match connections to a
+// given action.
+type TCPRouteMatch struct {
+	// SNI matches a route based on TLS server name indication.
+	//
+	// +optional
+	SNI *string `json:"sni,omitempty" protobuf:"bytes,1,opt,name=sni"`
```

Comment:
For TLS, the SNI name has to already be specified on the listener. What's the use case for specifying the name again here? If there is a use case for matching the name here, does that also apply to the port?

Reply:
Another TLS-specific match option could be the ALPN protocol, which isn't specified on the Listener.

Reply:
@jpeach This seems like the tls passthrough case where the SNI is sniffed by reading the TLS client hello. The gateway seems to allow multiple SNIs, but the specific SNIs route to specific backend services via the TCPRoute object? If unless the port is specified, maybe it refers to all port match for the backend service?

Reply:
> 1. We should add example for TCPRoute inside the `examples` dir as part of this PR or a follow up PR.
> 2. SNI inside TCPRoute is going to be contentious. We should consider merging this without the SNI part and then discuss the SNI part in a follow up PR.

I agree with both :)